File: source/javascript/dfs.js
// Depth-First Search Traversal
function dfs(adj, start) {
    const visited = new Array(adj.length).fill(false);
    const order = [];
    const stack = [start];

    while (stack.length > 0) {
        const u = stack.pop();
        if (visited[u]) continue;
        visited[u] = true;
        order.push(u);

        const neighbors = adj[u];
        for (let i = neighbors.length - 1; i >= 0; i--) {
            if (!visited[neighbors[i]]) stack.push(neighbors[i]);
        }
    }
    return order;
}

const adj = [[1, 2], [3, 4], [5], [], [5], []];
const result = dfs(adj, 0);
console.log(`dfs: ${result.join(" ")}`);

File: source/rust/dfs.rs
fn dfs(adj: &Vec<Vec<usize>>, start: usize) -> Vec<usize> {
    let n = adj.len();
    let mut visited = vec![false; n];
    let mut order = Vec::new();
    let mut stack = vec![start];

    while let Some(u) = stack.pop() {
        if visited[u] { continue; }
        visited[u] = true;
        order.push(u);
        for &v in adj[u].iter().rev() {
            if !visited[v] {
                stack.push(v);
            }
        }
    }
    order
}

fn main() {
    let adj = vec![
        vec![1, 2],
        vec![3, 4],
        vec![5],
        vec![],
        vec![5],
        vec![],
    ];
    let result = dfs(&adj, 0);
    let s: Vec<String> = result.iter().map(|x| x.to_string()).collect();
    println!("dfs: {}", s.join(" "));
}
